File: src/main/java/net/lomeli/lomlib/util/ByteUtil.java
package net.lomeli.lomlib.util;

import io.netty.buffer.ByteBuf;

import java.util.HashMap;
import java.util.Map;

import net.minecraft.util.BlockPos;

/**
 * Additional utils for ByteBuf
 */
public class ByteUtil {

    public static void writeIntArray(ByteBuf buffer, int... array) {
        buffer.writeInt(array.length);
        for (int i = 0; i < array.length; i++) {
            buffer.writeInt(array[i]);
        }
    }

    public static int[] readIntArray(ByteBuf buffer) {
        int[] returnArray = null;
        int size = buffer.readInt();
        if (size > 0) {
            returnArray = new int[size];
            for (int i = 0; i < size; i++) {
                returnArray[i] = buffer.readInt();
            }
        }
        return returnArray;
    }

    public static void writeIntMap(ByteBuf buffer, HashMap<Integer, Integer> map) {
        buffer.writeInt(map.size());
        for (Map.Entry<Integer, Integer> mapEntry : map.entrySet()) {
            int key = mapEntry.getKey();
            int value = mapEntry.getValue();
            buffer.writeInt(key);
            buffer.writeInt(value);
        }
    }

    public static HashMap<Integer, Integer> readIntMap(ByteBuf buffer) {
        HashMap<Integer, Integer> returnMap = new HashMap<Integer, Integer>();
        int size = buffer.readInt();
        if (size > 0) {
            for (int i = 0; i < size; i++) {
                int key = buffer.readInt();
                int value = buffer.readInt();
                returnMap.put(key, value);
            }
        }
        return returnMap;
    }

    public static void writeBlockPos(ByteBuf buffer, BlockPos pos) {
        buffer.writeInt(pos.getX());
        buffer.writeInt(pos.getY());
        buffer.writeInt(pos.getZ());
    }

    public static BlockPos readBlockPos(ByteBuf buffer) {
        int x = buffer.readInt();
        int y = buffer.readInt();
        int z = buffer.readInt();
        return new BlockPos(x, y, z);
    }
}

File: src/main/kotlin/net/lomeli/lomlib/util/forge/ByteUtil.kt
package net.lomeli.lomlib.util.forge

import io.netty.buffer.ByteBuf

import java.util.HashMap

import net.minecraft.util.math.BlockPos

/**
 * Additional utils for ByteBuf
 */
object ByteUtil {

    fun writeIntArray(buffer: ByteBuf, vararg array: Int) {
        buffer.writeInt(array.size)
        for (i in array) {
            buffer.writeInt(array[i])
        }
    }

    fun readIntArray(buffer: ByteBuf): IntArray? {
        var returnArray: IntArray? = null
        val size = buffer.readInt()
        if (size > 0) {
            returnArray = IntArray(size)
            for (i in 0..size - 1) {
                returnArray[i] = buffer.readInt()
            }
        }
        return returnArray
    }

    fun writeIntMap(buffer: ByteBuf, map: HashMap<Int, Int>) {
        buffer.writeInt(map.size)
        for (entry in map.entries) {
            buffer.writeInt(entry.key)
            buffer.writeInt(entry.value)
        }
    }

    fun readIntMap(buffer: ByteBuf): HashMap<Int, Int> {
        val returnMap = HashMap<Int, Int>()
        val size = buffer.readInt()
        if (size > 0) {
            for (i in 0..size - 1) {
                val key = buffer.readInt()
                val value = buffer.readInt()
                returnMap.put(key, value)
            }
        }
        return returnMap
    }

    fun writeBlockPos(buffer: ByteBuf, pos: BlockPos) {
        buffer.writeLong(pos.toLong())
    }

    fun readBlockPos(buffer: ByteBuf): BlockPos {
        val posLong = buffer.readLong()
        return BlockPos.fromLong(posLong)
    }
}